client/exec.go

```diff
@@ -0,0 +1,312 @@
+// Copyright (c) 2021 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package client
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"time"
+
+	"github.com/gorilla/websocket"
+
+	"github.com/canonical/pebble/internal/wsutil"
+)
+
+// ExecOptions are the options for an Exec call.
+type ExecOptions struct {
+	// Required: command and arguments (first element is the executable).
+	Command []string
+
+	// Optional environment variables.
+	Environment map[string]string
+
+	// Optional working directory (default is $HOME or "/" if $HOME not set).
+	WorkingDir string
+
+	// Optional timeout for the command execution, after which the process
+	// will be terminated. If zero, no timeout applies.
+	Timeout time.Duration
+
+	// Optional user ID and group ID for the process to run as.
+	UserID  *int
+	User    string
+	GroupID *int
+	Group   string
```

Comment:
As we know, the int pointers are a bit of an awkward API.  We use that elsewhere in an internal library, and although I'm not super happy with it either, I didn't complain too much at the time because it's internal. But this is the public client library that we'll be handing to people for interacting with Pebble, so it feels like we ought to have something we'd be proud of.  

How about `ChangeUser` and `ChangeGroup` booleans? Any other suggestions?

Reply:
I kinda disagree on this one. I acknowledge pointer-based fields *are* a bit awkward -- I even opened a [golang/go issue](https://github.com/golang/go/issues/22647) to add syntax to the language to make it easier to create pointers in an expression; that issue was eventually closed, but Rob Pike has [since taken up](https://github.com/golang/go/issues/45624) the idea again, so we'll see.

That said, they have a few things going for them: 1) They use only basic Go types, and don't have the ergonomic problems of the `ChangeUser` additional boolean and 2) This pointer-to-mean-optional is a very common practice, especially in client libraries, such as the AWS SDK, Protobufs, etc. From my observation of Go practice (not a scientific measurement), it's much more common than an additional bool or an "optional" value type (that's used in `database/sql` for `NullInt32` and the like, but I haven't seen it widely used).

Alternatives are ([example of all three variants here](https://play.golang.org/p/RP4P4ATcvQA)):

1. Leave as is but have a `client.IntPtr` or `client.Int` helper so it's easier to create a pointer from a user ID literal int (that's what the likes of the AWS SDK does). However, my guess is that the user ID will usually be in a variable, so you can just use `&` on an existing variable.
2. Add a `type OptionalInt { Value int; IsSet bool }` with a helper `client.Int()` that returns an is-set `OptionalInt` from a plain int, and make the `UserID` and `GroupID` fields of type `OptionalInt`. 
3. Your suggestion of adding `ChangeUser` and `ChangeGroup` bools alongside the struct fields. I don't like this much, as it's too easy to get wrong by setting the `UserID` but forgetting to set `ChangeUser`.

I think we should leave as is. We can always add a `client.IntPtr` helper later if it's an issue, or maybe Go will get the `&int(42)` or `new(42)` syntax at some point via https://github.com/golang/go/issues/45624.

Reply:
not sure how valid this is, but we could possibly just have User/Group where the string can container a id number or name.

Reply:
Huh, interesting thought. Incidentally, that's what I've done for the Pebble CLI: there's only `--user` and `--group` args, and if it is all digits, it sends the integer ID. Also, it looks like all-digit usernames are invalid, so that works:

```
$ useradd 123456
useradd: invalid user name '123456'
```

It does make the API a bit more awkward when actually using uids/gids, but I guess it's only a `strconv.Itoa(uid)` away. There's also precedent for uids/gids being strings in [`os/user`](https://pkg.go.dev/os/user). Though I think there it's more because on Windows its an SID (kind of a UUID-like string).

In any case, I'd be open to using strings, but I'll see what @niemeyer thinks.

Reply:
Also, for what it's worth, we already have the int UserID / string User in both the exec and files JSON APIs (though this is in the Go client, and we don't have a Go files client yet, so we could do something different).

Reply:
> I kinda disagree on this one. I acknowledge pointer-based fields are a bit awkward

Looks like you actually agree then :smile:

> 1. Leave as is but have a client.IntPtr or client.Int
> 2. Add a type OptionalInt { Value int; IsSet bool }

Neither of these feel better than just having a boolean saying "do this". Indeed, the second proposal is doing pretty much that, except at a more fine grained level. If that would be okay, I cannot see why something like SetUser/ChangeUser/Whatever would not be.

With that said, let's look for the middle ground. There must be a nice way to address this that is clean and elegant.

> not sure how valid this is, but we could possibly just have User/Group where the string can container a id number or name.

We discussed exactly that design before in a slightly different context, but still inside the Pebble / Operator Framework. We ended up backtracking from that design as we realized that it makes it impossible to provide both a username and a user ID, and that may be relevant in cases where we're asking for something to be done under a given username, but with a fallback in case that user does not exist.

So here is another proposal: let's just use an interface{} as the type, so that we can distinguish nil from 0. It's similar to using a pointer, but doesn't have the awkwardness of having to do this every time:
```
var userID *int
if setUserID {
    var i int = 0
    userID = &i
}
```

> or maybe Go will get the &int(42) or new(42) syntax at some point via

Perhaps it's just me, but this continues to taste like a hack to be honest. We really don't care about the fact that this is a pointer in terms of memory management, or in terms of it being a reference to some externally held data, which tends to be the actual reason why we have pointers. Instead, we're using a pointer because we want to make it optional while at the same time we want zero to be a valid value, which happens to be the default initialization for Go struct values.

So isn't it curious that we make it _awkward_ to make it _nice_? :)

That's why I'd rather design the API in a way that feels natural and consistent, instead of just taking that route.

Reply:
> So here is another proposal: let's just use an interface{} as the type, so that we can distinguish nil from 0. It's similar to using a pointer, but doesn't have the awkwardness of having to do this every time:

Oooh, I really don't like that. Now we're writing dynamically typed code (Python in Go!). The compiler wouldn't be able to check its type at all, so they'd only find out if they'd passed the wrong type at runtime (panic or error). And it'd be a fairly easy mistake to make to pass a string, because the Go [`user`](https://pkg.go.dev/os/user#User) package uses strings for this (to be cross-platform).

>     var userID *int
>     if setUserID {
>         var i int = 0
>         userID = &i
>     }

The thing is, most uses of this will be getting the user ID from somewhere else, not from an integer literal, so the above awkwardness would probably be rare. For example, in the one existing place we use the client (in the CLI in `cmd_exec.go`) it does exactly this: the value comes from a command line argument, which is already a `*int`, so it's just `UserID: cmd.UserID`. In other cases it might be `UserID: &userID`.

Another possibility is just to make 0 (root user) illegal. This might be okay, because if Pebble as running as root (true in a K8s context), you won't need to set the user/group, and if Pebble is not running as root, it will be a permission denied error to specify 0 (root) anyway. The exception would be if a user has `sudo` privileges. We could even have `UserRoot bool` and `GroupRoot bool` fields to specify that case: similar to your first suggestion of `ChangeUser bool`, but you only specify *either* the `UserID int` or the `UserRoot bool`. I prefer those over `interface{}`, but don't love either of them.

> We really don't care about the fact that this is a pointer in terms of memory management, or in terms of it being a reference to some externally held data, which tends to be the actual reason why we have pointers.

I hear what you're saying, and this is the *primary* use of pointers in Go, but using pointers to mean optional or "possibly not set" is very common in Go, just because pointers have a type but are nilable. For example, [`BranchStmt`](https://pkg.go.dev/go/ast#BranchStmt) in the `ast` package has an optional `Label` field done as a pointer. Hashicorp's Nomad library [uses it](https://github.com/hashicorp/nomad/blob/d486e1d500f7bc135d4e4f2d602acd434a37f8bd/nomad/structs/structs.go#L5349-L5350) for optional fields in their API structs (a similar use case as ours here). It's used for "might not be set" fields [in JSON](https://willnorris.com/2014/05/go-rest-apis-and-pointers/). And so on.

In summary, I think this is idiomatic Go, and won't be awkward in practice because using integer literals for uid/gid will be rare.

Reply:
> Oooh, I really don't like that. Now we're writing dynamically typed code (Python in Go!). 

More like Go in Go. Does this ring any bells:

```
Funcs map[string]interface{}
```
;-)

> Another possibility is just to make 0 (root user) illegal.
(...)
> it will be a permission denied error to specify 0 (root) anyway

It's often not a good idea to make something impossible by design because we have a soft feeling that we might not need it right now. The desire of running a command as root is far from special, and we may want to support that in Pebble properly.

> We could even have UserRoot bool and GroupRoot bool fields to specify that case

I'm trying to find a middle ground, but I'm failing to follow along your line of design. You were unhappy with having a boolean to indicate the desire to change the user, but this suggestion feels very much in that direction, but significantly more special and awkward in my subjective view.

> In summary, I think this is idiomatic Go

Repeating myself: yes, some people do it, but it doesn't stop it being ugly and inconvenient.
```
var userID *int
if setUserID {
    var i int = 0
    userID = &i
}
```
I'm trying to avoid this, and it seems easy to do so. Let's catch on a video call and find something we both agree with.

Reply:
We discussed via video, and came to the agreement that while `UserID *int` is a bit awkward, it's less bad than any of the alternatives we can think of, so we're going to stick with the status quo.